vernac/declare.ml

```diff
@@ -1958,7 +1958,7 @@ let build_by_tactic env ~uctx ~poly ~typ tac =
   let sigma = Evd.from_ctx uctx in
   let ce, status, sigma = build_constant_by_tactic ~name ~sigma ~sign ~poly typ tac in
   let uctx = Evd.evar_universe_context sigma in
-  let cb, uctx = inline_private_constants ~uctx env ce in
+  let cb, _uctx = inline_private_constants ~uctx env ce in
```

Comment:
Maybe what happens is that in the build_by_tactic / solve_by_tac / Solve Obligations and automatic obligation solving code path we never reset the state to before the tactic ran, so the side effects are still in the global env unlike the Next Obligation ... Qed path. Which would explain why we need to treat universes differently.

Reply:
indeed side effects even leak out:
~~~coq

Require Import Program.Tactics.
Set Printing Universes.

Local Obligation Tactic := idtac.

Polymorphic Program Definition foo : nat := _.
Solve Obligations with (try ((abstract exact 0); gfail)).

Print temporary_proof2_subproof.
(* (no body) : nat *)
~~~
(no body because of https://github.com/coq/coq/issues/13324)